### PR TITLE
Cmake improvements

### PR DESCRIPTION
**Pull request by mtaillefumier**
very simple changes

### CODE REVIEW

**Comment by oschuett** [comment]
Going forward, I think our next goal should be to add a test to prevent backsliding, e.g. when new files are added.

We could start small, ie. with the equivalent of [Dockerfile.test_minimal](https://github.com/cp2k/cp2k/blob/master/tools/docker/Dockerfile.test_minimal), and then extend the test's coverage over time.

This is what I have so far:
```
source /opt/cp2k-toolchain/install/setup
export PKG_CONFIG_PATH=/opt/cp2k-toolchain/install/openblas-0.3.21/lib/pkgconfig
cmake -DCP2K_USE_LIBXSMM=NO -DSCALAPACK_ROOT=/opt/cp2k-toolchain/install/scalapack-2.1.0 ..
```

Unfortunately, it fails with the following message:
```
 Could not find a package configuration file provided by "DBCSR" (requested
  version 2.2.0) with any of the following names:

    DBCSRConfig.cmake
    dbcsr-config.cmake
```

Maybe we could use this PR to develop such a minimal test? I would setup the CI and then you could fill in the right commands for building CP2K via cmake on top of the toolchain. What do you think?

**Comment by mtaillefumier** [comment]
> Going forward, I think our next goal should be to add a test to prevent backsliding, e.g. when new files are added.

I agree with you. If I do not follow the master branch for more than a week then it is very likely that a new file will be added or removed. 
 
> We could start small, ie. with the equivalent of [Dockerfile.test_minimal](https://github.com/cp2k/cp2k/blob/master/tools/docker/Dockerfile.test_minimal), and then extend the test's coverage over time.
> 
> This is what I have so far:
> 
> ```
> source /opt/cp2k-toolchain/install/setup
> export PKG_CONFIG_PATH=/opt/cp2k-toolchain/install/openblas-0.3.21/lib/pkgconfig
> cmake -DCP2K_USE_LIBXSMM=NO -DSCALAPACK_ROOT=/opt/cp2k-toolchain/install/scalapack-2.1.0 ..
> ```
> 
> Unfortunately, it fails with the following message:
> 
> ```
>  Could not find a package configuration file provided by "DBCSR" (requested
>   version 2.2.0) with any of the following names:
> 
>     DBCSRConfig.cmake
>     dbcsr-config.cmake
> ```
> 
> Maybe we could use this PR to develop such a minimal test? I would setup the CI and then you could fill in the right commands for building CP2K via cmake on top of the toolchain. What do you think?

Let's try to be constructive indeed and use this PR to include the minimal test. Could you make a separate PR for the minimal test so I can rebase on it ? Second option, just make a PR on my branch. Choose the option that you think is better.

The error message just states that dbcsr is nowhere to be found. I will take care of this for the CI. Alternatively you can add `-DDBCSR_ROOT=/opt/cp2k-toolchain/install/dbcsr` 

I usually build and install all cp2k dependencies in the same directory. so doing `cmake -DCMAKE_INSTALL_PREFIX=/mypath ..` just works.

I will have a look at the dockerfile as well.

**Comment by oschuett** [comment]
I've merged my #2389. So, now you can rebase this PR and start working on the [Dockerfile](https://github.com/cp2k/cp2k/blob/d7412ee8675cb76473b00ae3e51e7018021298a2/tools/docker/generate_dockerfiles.py#L368).

Once the build works we can merge this PR. Then I'll followup with a refactoring of `do_regtest.py`.

**Comment by mtaillefumier** [comment]
Working on it actually but my laptop is slow at generating the docker container....

**Comment by oschuett** [comment]
It's very good that you are moving the DBCSR build into the toolchain (#351). However, it's a bit of a large change on its own. Perhaps this warrants a separate PR?

**Comment by mtaillefumier** [comment]
I am stuck with an issue I have no definite solution to yet (order of library does not always seem to be constant). 

DBCSR is not included in the toolchain script because I do not want to break cp2k build system. I just wrote the dbcsr script to ease my life but I do not really want to include it in the toolchain at the moment. Basically it is a place holder/time saver for the time being.

**Comment by mtaillefumier** [comment]
I found out my mistake. quite simple actually. @oschuett would it be possible to trigger the cmake docker build. It works on my machine but I want to check on the hardware used by the CI.

- make pretty will fail. 
- Do we want the docker build to test the most basic setup or include the reasonable dependencies (libxc, libxsmm, libint2)?
- I will find a way to build dbcsr without using the script. It was a place holder til I fix the cmake build system

**Comment by oschuett** [comment]
> I found out my mistake.

Great :-)

> @oschuett would it be possible to trigger the cmake docker build.

Done! Btw, if you only add new commits, ie. no push force, then the CI will automatically re-run the checks from the previous commit.

> Do we want the docker build to test the most basic setup or include the reasonable dependencies (libxc, libxsmm, libint2)?

Please include as many dependencies as possible. The idea is to use this separate cmake tests as an incubator before I can migrate the other dashboard tests.

> I will find a way to build dbcsr without using the script.

Sounds good. I might take a stab at #351 myself in the coming weeks. I've been meaning to do it for a while.

**Comment by mtaillefumier** [comment]
The script build dbcsr is actually a modified version of some other packages provided by the toolchain. I hacked my way to avoid solving an issue that is docker specific in that case. 
I will include as many dependencies as possible.

**Comment by mtaillefumier** [comment]
we need to update scalapack actually. Curiously scalapack does not build within the docker environment.

File: tools/toolchain/scripts/stage4/install_scalapack.sh [hunk 1]
```diff
@@ -39,26 +39,29 @@ case "$with_scalapack" in
       if [ -f ${scalapack_pkg} ]; then
         echo "${scalapack_pkg} is found"
       else
-        download_pkg_from_cp2k_org "${scalapack_sha256}" "${scalapack_pkg}"
+	wget https://github.com/Reference-ScaLAPACK/scalapack/archive/refs/tags/v2.2.1.tar.gz -O "${scalapack_pkg}"
+        #download_pkg_from_cp2k_org "${scalapack_sha256}" "${scalapack_pkg}"
```
Comment (oschuett) [inline tools/toolchain/scripts/stage4/install_scalapack.sh]: I've uploaded [scalapack-2.2.1.tar.gz](https://www.cp2k.org/static/downloads/scalapack-2.2.1.tar.gz) to our server. So you can now use `download_pkg_from_cp2k_org` again.

File: tools/toolchain/scripts/stage4/install_scalapack.sh [hunk 2]
```diff
@@ -39,26 +39,29 @@ case "$with_scalapack" in
...
       fi
       echo "Installing from scratch into ${pkg_install_dir}"
       [ -d scalapack-${scalapack_ver} ] && rm -rf scalapack-${scalapack_ver}
       tar -xzf ${scalapack_pkg}
 
       pushd "scalapack-${scalapack_ver}" > /dev/null
-      for patch in "${patches[@]}"; do
-        patch -p1 < "${patch}" >> patch.log 2>&1
-      done
+      #for patch in "${patches[@]}"; do
```
Comment (oschuett) [inline tools/toolchain/scripts/stage4/install_scalapack.sh]: Feel free to remove the patching code entirely - including `pushd`,`popd`, and [scalapack-2.1.0-gcc10.patch](https://github.com/cp2k/cp2k/blob/master/tools/toolchain/scripts/stage4/scalapack-2.1.0-gcc10.patch)

**Comment by oschuett** [comment]
It seems the cmake build was successful :-)
So just move your code into [generate_dockerfiles.py](https://github.com/cp2k/cp2k/blob/master/tools/docker/generate_dockerfiles.py#L375) and then it's ready to be merged.

**Comment by mtaillefumier** [comment]
I wanted to add more dependencies to the mix but so far if the first step is to be able to test compilation then I think it is already good. 
I can take care of the dbcsr script but I won"t call it from the install_cp2k_toolchain.sh for the time being.

**Comment by mtaillefumier** [comment]
Yeah sorry for the ugly hack in the scalapack package. I will clean it. Do we want to update it ?

**Comment by oschuett** [comment]
> if the first step is to be able to test compilation then I think it is already good.

Perfect, then let's merge it. Afterwards I can setup the daily dashboard run and we can stop worrying about backsliding.

> I can take care of the dbcsr script but I won"t call it from the install_cp2k_toolchain.sh for the time being.

You can put your helper scripts in [tools/docker/scripts](https://github.com/cp2k/cp2k/tree/master/tools/docker/scripts).

> Yeah sorry for the ugly hack in the scalapack package. I will clean it. Do we want to update it ?

Yes, I'd upgrade ScaLAPACK in any case simply to get ride of that large patch file.

**Comment by mtaillefumier** [comment]
> > if the first step is to be able to test compilation then I think it is already good.
> 
> Perfect, then let's merge it. Afterwards I can setup the daily dashboard run and we can stop worrying about backsliding.
> 
> > I can take care of the dbcsr script but I won"t call it from the install_cp2k_toolchain.sh for the time being.
> 
> You can put your helper scripts in [tools/docker/scripts](https://github.com/cp2k/cp2k/tree/master/tools/docker/scripts).
> 
> > Yeah sorry for the ugly hack in the scalapack package. I will clean it. Do we want to update it ?
> 
> Yes, I'd upgrade ScaLAPACK in any case simply to get ride of that large patch file.

OK. Let me do the clean up and commit the changes.

File: 0001-Update-scalapack.patch [hunk 3]
```diff
@@ -0,0 +1,5680 @@
+From 402145694a9e0fc578abf2d8f47c3dd939688c92 Mon Sep 17 00:00:00 2001
```
Comment (oschuett) [inline 0001-Update-scalapack.patch]: Seems we got a stowaway :-)

**Comment by mtaillefumier** [comment]
Oh the little bugger....

**Comment by oschuett** [comment]
To make the Python test happy you simply need to run the `generate_dockerfiles.py` script.

**Comment by abussy** [comment]
It looks like this PR causes troubles with the installation of ScalaPACK with older version of gfortran. On a local machine, using the system gfortran compiler (version 7.5.0), I get the following error from the toolchain script:

```
==================== Installing ScaLAPACK ====================
scalapack-2.2.1.tgz: OK
Checksum of scalapack-2.2.1.tgz Ok
Installing from scratch into /data/aubuss/cp2k/tools/toolchain/install/scalapack-2.2.1
-- The C compiler identification is GNU 7.5.0
-- The Fortran compiler identification is GNU 7.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - failed
-- Check for working Fortran compiler: /usr/bin/gfortran
-- Check for working Fortran compiler: /usr/bin/gfortran - broken
CMake Error at /data/aubuss/cp2k/tools/toolchain/install/cmake-3.24.2/share/cmake-3.24/Modules/CMakeTestFortranCompiler.cmake:61 (message):
  The Fortran compiler

    "/usr/bin/gfortran"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_d34f3/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_d34f3.dir/build.make CMakeFiles/cmTC_d34f3.dir/build
    gmake[1]: Entering directory '/data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp'
    Building Fortran object CMakeFiles/cmTC_d34f3.dir/testFortranCompiler.f.o
    /usr/bin/gfortran   -fallow-argument-mismatch  -c /data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_d34f3.dir/testFortranCompiler.f.o
    gfortran: error: unrecognized command line option ‘-fallow-argument-mismatch’; did you mean ‘-Wno-argument-mismatch’?
    gmake[1]: *** [CMakeFiles/cmTC_d34f3.dir/build.make:78: CMakeFiles/cmTC_d34f3.dir/testFortranCompiler.f.o] Error 1
    gmake[1]: Leaving directory '/data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_d34f3/fast] Error 2

```

It looks like the culprit is the `-fallow-argument-mismatch` compilation flag, which seems to only exist for later versions of gfortran. Indeed, if I also install GCC with the toolchain (version 12.2.0), everything runs smoothly.

I am not sure what the policy is with respect to GCC versions, but I thought you should be made aware of this.

**Comment by mkrack** [comment]
A check like [this](https://github.com/cp2k/cp2k/blob/master/arch/Linux-gnu-x86_64.psmp#L386-L388) is needed.

**Comment by mtaillefumier** [comment]
> It looks like this PR causes troubles with the installation of ScalaPACK with older version of gfortran. On a local machine, using the system gfortran compiler (version 7.5.0), I get the following error from the toolchain script:
> 
> ```
> ==================== Installing ScaLAPACK ====================
> scalapack-2.2.1.tgz: OK
> Checksum of scalapack-2.2.1.tgz Ok
> Installing from scratch into /data/aubuss/cp2k/tools/toolchain/install/scalapack-2.2.1
> -- The C compiler identification is GNU 7.5.0
> -- The Fortran compiler identification is GNU 7.5.0
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Check for working C compiler: /usr/bin/gcc - skipped
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Detecting Fortran compiler ABI info
> -- Detecting Fortran compiler ABI info - failed
> -- Check for working Fortran compiler: /usr/bin/gfortran
> -- Check for working Fortran compiler: /usr/bin/gfortran - broken
> CMake Error at /data/aubuss/cp2k/tools/toolchain/install/cmake-3.24.2/share/cmake-3.24/Modules/CMakeTestFortranCompiler.cmake:61 (message):
>   The Fortran compiler
> 
>     "/usr/bin/gfortran"
> 
>   is not able to compile a simple test program.
> 
>   It fails with the following output:
> 
>     Change Dir: /data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp
>     
>     Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_d34f3/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_d34f3.dir/build.make CMakeFiles/cmTC_d34f3.dir/build
>     gmake[1]: Entering directory '/data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp'
>     Building Fortran object CMakeFiles/cmTC_d34f3.dir/testFortranCompiler.f.o
>     /usr/bin/gfortran   -fallow-argument-mismatch  -c /data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_d34f3.dir/testFortranCompiler.f.o
>     gfortran: error: unrecognized command line option ‘-fallow-argument-mismatch’; did you mean ‘-Wno-argument-mismatch’?
>     gmake[1]: *** [CMakeFiles/cmTC_d34f3.dir/build.make:78: CMakeFiles/cmTC_d34f3.dir/testFortranCompiler.f.o] Error 1
>     gmake[1]: Leaving directory '/data/aubuss/cp2k/tools/toolchain/build/scalapack-2.2.1/build/CMakeFiles/CMakeTmp'
>     gmake: *** [Makefile:127: cmTC_d34f3/fast] Error 2
> ```
> 
> It looks like the culprit is the `-fallow-argument-mismatch` compilation flag, which seems to only exist for later versions of gfortran. Indeed, if I also install GCC with the toolchain (version 12.2.0), everything runs smoothly.
> 
> I am not sure what the policy is with respect to GCC versions, but I thought you should be made aware of this.

I can not and will not cover all versions of gcc as the issue is not with the compiler but scalapack code. CP2K code has the same issue here and there. gcc changed the policy about this with few versions ago (I think gcc 10). There are several options to go around the issue

1. update the gcc version (which is probably the best)
2. do what @mkrack has suggested (but adapt it to the toolchain)
3. fix scalapack code or build system but the issue will still be present afterwards when cp2k is compiled as the option is automatically added to the FFLAGS by the toolchain.

**Comment by alazzaro** [comment]
I would say that patching SCALAPACK cmake (we had something similar in the past with tools/toolchain/scripts/stage4/scalapack-2.1.0-gcc10.patch, which was deleted) is something possible. At the end the patch is something like what we have in DBCSR:

https://github.com/cp2k/dbcsr/blob/b7a46715d391dc8ea604b512caf25dcb7f484d2b/cmake/CompilerConfiguration.cmake#L3

**Comment by abussy** [comment]
Thanks for the inputs. PR #2439 should fix this.

**Pull request merged**